### PR TITLE
CASMHMS-5571 Improve hms-bss chart tests

### PR DESCRIPTION
**Pull request by mtupitsyn**
## Summary and Scope

Minor improvements to helm-bss-charts tests.

## Issues and Related PRs

* Resolves [CASMHMS-5571](https://jira-pro.its.hpecorp.net:8443/browse/CASMHMS-5571)
* Relates to [CASMINST-4067](https://jira-pro.its.hpecorp.net:8443/browse/CASMINST-4067)

## Testing

Tested on:

  * Virtual Shasta

Test description:

Ensured proper test passing (test for false negative) and proper recognition of test failures (test for false positive) using `helm test` command against chart, deployed to vshasta.

## Risks and Mitigations

Low - test changes only for a chart version, not yet accepted by CSM build.

## Pull Request Checklist

- [x] Version number(s) incremented, if applicable
- [X] Copyrights updated
- [x] License file intact
- [x] Target branch correct
- [x] CHANGELOG.md updated
- [x] Testing is appropriate and complete, if applicable

### CODE REVIEW

File: charts/v2.1/cray-hms-bss/templates/tests/test-functional.yaml [hunk 1]
```diff
@@ -12,9 +35,12 @@ metadata:
     app.kubernetes.io/name: "{{ .Release.Name }}-test-functional"
 
 spec:
+  backoffLimit: 0
```
Comment (mtupitsyn) [inline charts/v2.1/cray-hms-bss/templates/tests/test-functional.yaml:38]: This ensures that test job spawns only one pod and returns the result afterwards. Default behavior for Kubernetes jobs is to span up to 6 consecutive pods in case of failure. Each pod is spawned with exponential delay (10s, 20s, etc). This usually takes more then 5 minutes, so `helm test` command times out (if run with with default timeout) instead of returning negative test result.

File: charts/v2.1/cray-hms-bss/templates/tests/test-functional.yaml [hunk 2]
```diff
@@ -30,4 +56,4 @@ spec:
           image: "{{ .Values.tests.image.repository }}:{{ .Values.global.testVersion }}"
           imagePullPolicy: "{{ .Values.tests.image.pullPolicy }}"
           command: ["/bin/sh", "-c"]
-          args: [ "until curl --head localhost:15000 ; do echo Waiting for Sidecar; sleep 3 ; done ; echo Sidecar available; entrypoint.sh functional -c /src/libs/tavern_global_config.yaml -p /src/app"]
+          args: [ "entrypoint.sh functional -c /src/libs/tavern_global_config.yaml -p /src/app"]
```
Comment (mtupitsyn) [inline charts/v2.1/cray-hms-bss/templates/tests/test-functional.yaml:59]: Since istio 1.7, the same functionality is achieved with configuration option `holdApplicationUntilProxyStarts`, provided through pod annotation.

**Review by schooler-hpe — APPROVED** [review]

**Comment by schooler-hpe** [comment]
Nice improvements! I'm going to make these same changes to all of the HMS Charts with Helm Tests.

Looks like your PR is blocked due to Synk finding vulnerabilities in the alpine:3.15 base image that's being pulled in.

**Comment by mtupitsyn** [comment]
> Looks like your PR is blocked due to Synk finding vulnerabilities in the alpine:3.15 base image that's being pulled in.

That's correct. In fact, we do periodic rebuild of images built the old way (in jenkins) - we locate Jenkins jobs by image tags and re-trigger them. Since HMS team has separate workflow for docker image builds, it is something HMS team will need to take care of. Maybe just add a rebuild schedule to github build workflows.

Another issue with HMS image build workflows - it looks like Snyk test is executed against the image stored in Artifactory, and not against a new image which was just built from dev branch. Note, all lint/test/scan activities are happening in a different job from a build, which means they are on different runner and different instance of docker daemon... So it just tested `artifactory.algol60.net/csm-docker/stable/cray-bss-test:1.16.0` built long time ago from main branch. To actually test what you have built, you'll need to ensure that build and test is happening on the same runner (i.e. within the same job). Alternatively, you may want to produce dev tags (timestamped, with git SHA) from dev branches, push dev image to artifactory with this unique tag and test it on another runner. But this willalso require a cleanup job, otherwise artifactory will quickly fill up with dev images.

**Comment by mtupitsyn** [comment]
Would it be ok if I merge the PR (I can override scan-images result)?

**Review by nieuwsma — APPROVED** [review]

**Comment by schooler-hpe** [comment]
Thanks for the heads up about the Snyk testing in the image build workflows. I will pass that along to the team.

I'm fine with overriding the scan-images result and merging, but you should also update the following before doing so:
- Bump the Chart up a patch version (Z in X.Y.Z for this case)
- Update the cray-hms-bss.compatibility.yaml file with the new Chart version
- Update the changelog/v2.1.md file with your change

If you'd rather have me make those changes for you before merging that's no problem.

UPDATE: I didn't reload this page and missed that Andrew already merged your change. Nevermind!
